stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1056 @@
+            &parent_chain_tip.consensus_hash,
+            &block.header,
+            chain_tip_consensus_hash,
+            chain_tip_burn_header_hash,
+            chain_tip_burn_header_height,
+            chain_tip_burn_header_timestamp,
+            scheduled_miner_reward.as_ref(),
+            user_burns,
+            miner_payouts_opt,
+            &block_execution_cost,
+            &total_tenure_cost,
+            block_size,
+            applied_epoch_transition,
+            burn_stack_stx_ops,
+            burn_transfer_stx_ops,
+            burn_delegate_stx_ops,
+            affirmation_weight,
+        )
+        .expect("FATAL: failed to advance chain tip");
+
+        chainstate_tx.log_transactions_processed(&new_tip.index_block_hash(), &tx_receipts);
+
+        monitoring::set_last_block_transaction_count(block.txs.len() as u64);
+        monitoring::set_last_execution_cost_observed(&block_execution_cost, &block_limit);
+
+        // get previous burn block stats
+        let (parent_burn_block_hash, parent_burn_block_height, parent_burn_block_timestamp) =
+            if block.is_first_mined() {
+                (BurnchainHeaderHash([0; 32]), 0, 0)
+            } else {
+                match SortitionDB::get_block_snapshot_consensus(burn_dbconn, &parent_ch)? {
+                    Some(sn) => (
+                        sn.burn_header_hash,
+                        sn.block_height as u32,
+                        sn.burn_header_timestamp,
+                    ),
+                    None => {
+                        // shouldn't happen
+                        warn!(
+                            "CORRUPTION: block {}/{} does not correspond to a burn block",
+                            &parent_ch, &parent_block_hash
+                        );
+                        (BurnchainHeaderHash([0; 32]), 0, 0)
+                    }
+                }
+            };
+
+        let epoch_receipt = StacksEpochReceipt {
+            header: new_tip,
+            tx_receipts,
+            matured_rewards,
+            matured_rewards_info,
+            parent_microblocks_cost: ExecutionCost::zero(),
+            anchored_block_cost: block_execution_cost,
+            parent_burn_block_hash,
+            parent_burn_block_height,
+            parent_burn_block_timestamp,
+            evaluated_epoch,
+            epoch_transition: applied_epoch_transition,
+        };
+
+        Ok((epoch_receipt, clarity_commit))
+    }
+}
+
+impl StacksMessageCodec for NakamotoBlock {
+    fn consensus_serialize<W: std::io::Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.header)?;
+        write_next(fd, &self.txs)
+    }
+
+    fn consensus_deserialize<R: std::io::Read>(fd: &mut R) -> Result<Self, CodecError> {
+        let header: NakamotoBlockHeader = read_next(fd)?;
+        let txs: Vec<_> = {
+            let mut bound_read = BoundReader::from_reader(fd, MAX_MESSAGE_LEN as u64);
```

Comment:
Nit -- you should use a `bound_read` for decoding the `NakamotoBlockHeader` as well as the txs.  The header plus the txs should not exceed `MAX_MESSAGE_LEN` bytes.

Reply:
Updated